### Preparation - Importing all necessary modules

In [44]:
# import all required libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [45]:
# Display maximum rows and columns with pandas
pd.set_option('display.max_columns', 150)
pd.set_option('display.max_rows', 100)

### Data Collection - Loading the data sets

In [46]:
# Load the data set
df_for_loan = pd.read_csv("loan.csv")

In [47]:
df_for_loan.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-11   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-11  Charged Off   
2           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85               1                     NaN   
1           Apr-99               5                     NaN   
2           Nov-01               2                     NaN   
3           Feb-96               1                    35.0   
4           Jan-96               0                    38.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         3        0      13648     83.70%          9   
1                     NaN         3        0       1687      9.40%          4   
2                     NaN         2        0       2956     98.50%         10   
3                     NaN        10        0       5598        21%         37   
4                     NaN        15        0      27783     53.90%         38   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   f       0.00           0.00   5863.155187   
1                   f       0.00           0.00   1008.710000   
2                   f       0.00           0.00   3005.666844   
3                   f       0.00           0.00  12231.890000   
4                   f     524.06         524.06   3513.330000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16     

In [48]:
df_for_loan.columns

Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade',
       ...
       'num_tl_90g_dpd_24m', 'num_tl_op_past_12m', 'pct_tl_nvr_dlq',
       'percent_bc_gt_75', 'pub_rec_bankruptcies', 'tax_liens',
       'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit',
       'total_il_high_credit_limit'],
      dtype='object', length=111)

In [49]:
df_for_loan.describe()

id     member_id     loan_amnt   funded_amnt  \
count  3.971700e+04  3.971700e+04  39717.000000  39717.000000   
mean   6.831319e+05  8.504636e+05  11219.443815  10947.713196   
std    2.106941e+05  2.656783e+05   7456.670694   7187.238670   
min    5.473400e+04  7.069900e+04    500.000000    500.000000   
25%    5.162210e+05  6.667800e+05   5500.000000   5400.000000   
50%    6.656650e+05  8.508120e+05  10000.000000   9600.000000   
75%    8.377550e+05  1.047339e+06  15000.000000  15000.000000   
max    1.077501e+06  1.314167e+06  35000.000000  35000.000000   

       funded_amnt_inv   installment    annual_inc           dti  \
count     39717.000000  39717.000000  3.971700e+04  39717.000000   
mean      10397.448868    324.561922  6.896893e+04     13.315130   
std        7128.450439    208.874874  6.379377e+04      6.678594   
min           0.000000     15.690000  4.000000e+03      0.000000   
25%        5000.000000    167.020000  4.040400e+04      8.170000   
50%        8975.000000    280.220000  5.900000e+04     13.400000   
75%       14400.000000    430.780000  8.230000e+04     18.600000   
max       35000.000000   1305.190000  6.000000e+06     29.990000   

        delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
count  39717.000000    39717.000000            14035.000000   
mean       0.146512        0.869200               35.900962   
std        0.491812        1.070219               22.020060   
min        0.000000        0.000000                0.000000   
25%        0.000000        0.000000               18.000000   
50%        0.000000        1.000000               34.000000   
75%        0.000000        1.000000               52.000000   
max       11.000000        8.000000              120.000000   

       mths_since_last_record      open_acc       pub_rec      revol_bal  \
count             2786.000000  39717.000000  39717.000000   39717.000000   
mean                69.698134      9.294408      0.055065   13382.528086   
std                 43.822529      4.400282      0.237200   15885.016641   
min                  0.000000      2.000000      0.000000       0.000000   
25%                 22.000000      6.000000      0.000000    3703.000000   
50%                 90.000000      9.000000      0.000000    8850.000000   
75%                104.000000     12.000000      0.000000   17058.000000   
max                129.000000     44.000000      4.000000  149588.000000   

          total_acc     out_prncp  out_prncp_inv   total_pymnt  \
count  39717.000000  39717.000000   39717.000000  39717.000000   
mean      22.088828     51.227887      50.989768  12153.596544   
std       11.401709    375.172839     373.824457   9042.040766   
min        2.000000      0.000000       0.000000      0.000000   
25%       13.000000      0.000000       0.000000   5576.930000   
50%       20.000000      0.000000       0.000000   9899.640319   
75%       29.000000      0.000000       0.000000  16534.433040   
max       90.000000   6311.470000    6307.370000  58563.679930   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
count     39717.000000     39717.000000   39717.000000        39717.000000   
mean      11567.149118      9793.348813    2263.663172            1.363015   
std        8942.672613      7065.522127    2608.111964            7.289979   
min           0.000000         0.000000       0.000000            0.000000   
25%        5112.310000      4600.000000     662.180000            0.000000   
50%        9287.150000      8000.000000    1348.910000            0.000000   
75%       15798.810000     13653.260000    2833.400000            0.000000   
max       58563.680000     35000.020000   23563.680000          180.200000   

         recoveries  collection_recovery_fee  last_pymnt_amnt  \
count  39717.000000             39717.000000     39717.000000   
mean      95.221624                12.406112      2678.826162   
std      688.744771               148.671593      4447.136012   
min    

In [50]:
df_for_loan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 33.6+ MB


In [51]:
df_for_loan.shape

(39717, 111)

### Inpecting the loan data, it has been found that we have 39717 rows and 111 columns.

### Now we intend to find the missing values on the data set and count along with the percentage so we understand the data in a better way.

In [52]:
# Checking for the number of missing values on the dataset
missing_values_count = df_for_loan.isnull().sum()

# Print the count of missing values
print(missing_values_count)

id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
tax_liens                        39
tot_hi_cred_lim               39717
total_bal_ex_mort             39717
total_bc_limit                39717
total_il_high_credit_limit    39717
Length: 111, dtype: int64


In [53]:
# Checking the percentage of missing value on the dataset
missing_percentage = (missing_values_count / len(df_for_loan)) * 100

# Displaying the missing data count and percentage together
print(pd.DataFrame({'Total Missing': missing_values_count, 'Percentage': missing_percentage}).sort_values(by='Percentage', ascending=False))

                           Total Missing  Percentage
verification_status_joint          39717       100.0
annual_inc_joint                   39717       100.0
mo_sin_old_rev_tl_op               39717       100.0
mo_sin_old_il_acct                 39717       100.0
bc_util                            39717       100.0
...                                  ...         ...
delinq_amnt                            0         0.0
policy_code                            0         0.0
earliest_cr_line                       0         0.0
delinq_2yrs                            0         0.0
id                                     0         0.0

[111 rows x 2 columns]


### Data Cleaning - It has been found that we have greater number of null values in the Dataset. So, we will  remove the null values.

In [54]:
# Drop the columns with null value
df_for_loan.dropna(axis=1, how="any", subset=None, inplace=True)

df_for_loan.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade home_ownership  annual_inc  \
0   10.65%       162.87     B        B2           RENT     24000.0   
1   15.27%        59.83     C        C4           RENT     30000.0   
2   15.96%        84.33     C        C5           RENT     12252.0   
3   13.49%       339.31     C        C1           RENT     49200.0   
4   12.69%        67.79     B        B5           RENT     80000.0   

  verification_status issue_d  loan_status pymnt_plan  \
0            Verified  Dec-11   Fully Paid          n   
1     Source Verified  Dec-11  Charged Off          n   
2        Not Verified  Dec-11   Fully Paid          n   
3     Source Verified  Dec-11   Fully Paid          n   
4     Source Verified  Dec-11      Current          n   

                                                 url         purpose zip_code  \
0  https://lendingclub.com/browse/loanDetail.acti...     credit_card    860xx   
1  https://lendingclub.com/browse/loanDetail.acti...             car    309xx   
2  https://lendingclub.com/browse/loanDetail.acti...  small_business    606xx   
3  https://lendingclub.com/browse/loanDetail.acti...           other    917xx   
4  https://lendingclub.com/browse/loanDetail.acti...           other    972xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  \
0         AZ  27.65            0           Jan-85               1         3   
1         GA   1.00            0           Apr-99               5         3   
2         IL   8.72            0           Nov-01               2         2   
3         CA  20.00            0           Feb-96               1        10   
4         OR  17.94            0           Jan-96               0        15   

   pub_rec  revol_bal  total_acc initial_list_status  out_prncp  \
0        0      13648          9                   f       0.00   
1        0       1687          4                   f       0.00   
2        0       2956         10                   f       0.00   
3        0       5598         37                   f       0.00   
4        0      27783         38                   f     524.06   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   5863.155187          5833.84          5000.00   
1           0.00   1008.710000          1008.71           456.46   
2           0.00   3005.666844          3005.67          2400.00   
3           0.00  12231.890000         12231.89         10000.00   
4         524.06   3513.330000          3513.33          2475.94   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                0.00        0.00                     0.00   
1         435.17                0.00      117.08                     1.11   
2         605.67                0.00        0.00                     0.00   
3        2214.92               16.97        0.00                     0.00   
4        1037.39                0.00        0.00                     0.00   

   last_pymnt_amnt  policy_code application_type  acc_now_delinq  delinq_amnt  
0           171.62            1       INDIVIDUAL               0            0  
1           119.66            1       INDIVIDUAL               0            0  
2           649.91            1       INDIVIDUAL               0            0  
3           357.48            1       INDIVIDUAL               0            0  
4            67.79            1       INDIVIDUAL               0            0

In [55]:
# Checking and printing duplicated rows
print(f"Number of duplicate rows: {df.duplicated().sum()}")

Number of duplicate rows: 0


In [61]:
df_for_loan.shape

(39717, 43)

### Let us find out the loan status for fully accepted loan

In [62]:
df_for_loan['loan_status'].value_counts()

loan_status
Fully Paid     32950
Charged Off     5627
Current         1140
Name: count, dtype: int64

### Since we want to understand the problem areas leading to default. And the Current loans are ongoing and will not add any value to our analysis so we will drop them

In [63]:
# Exclude 'Current' loans from the dataset
df_for_loan = df_for_loan[df_for_loan['loan_status'] != 'Current']

# Verify the counts of 'loan_status'
print(df_for_loan['loan_status'].value_counts())

loan_status
Fully Paid     32950
Charged Off     5627
Name: count, dtype: int64


In [64]:
df_for_loan.shape

(38577, 43)

In [65]:
df_for_loan.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
5  1075269    1311441       5000         5000           5000.0   36 months   

  int_rate  installment grade sub_grade home_ownership  annual_inc  \
0   10.65%       162.87     B        B2           RENT     24000.0   
1   15.27%        59.83     C        C4           RENT     30000.0   
2   15.96%        84.33     C        C5           RENT     12252.0   
3   13.49%       339.31     C        C1           RENT     49200.0   
5    7.90%       156.46     A        A4           RENT     36000.0   

  verification_status issue_d  loan_status pymnt_plan  \
0            Verified  Dec-11   Fully Paid          n   
1     Source Verified  Dec-11  Charged Off          n   
2        Not Verified  Dec-11   Fully Paid          n   
3     Source Verified  Dec-11   Fully Paid          n   
5     Source Verified  Dec-11   Fully Paid          n   

                                                 url         purpose zip_code  \
0  https://lendingclub.com/browse/loanDetail.acti...     credit_card    860xx   
1  https://lendingclub.com/browse/loanDetail.acti...             car    309xx   
2  https://lendingclub.com/browse/loanDetail.acti...  small_business    606xx   
3  https://lendingclub.com/browse/loanDetail.acti...           other    917xx   
5  https://lendingclub.com/browse/loanDetail.acti...         wedding    852xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  \
0         AZ  27.65            0           Jan-85               1         3   
1         GA   1.00            0           Apr-99               5         3   
2         IL   8.72            0           Nov-01               2         2   
3         CA  20.00            0           Feb-96               1        10   
5         AZ  11.20            0           Nov-04               3         9   

   pub_rec  revol_bal  total_acc initial_list_status  out_prncp  \
0        0      13648          9                   f        0.0   
1        0       1687          4                   f        0.0   
2        0       2956         10                   f        0.0   
3        0       5598         37                   f        0.0   
5        0       7963         12                   f        0.0   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0            0.0   5863.155187          5833.84          5000.00   
1            0.0   1008.710000          1008.71           456.46   
2            0.0   3005.666844          3005.67          2400.00   
3            0.0  12231.890000         12231.89         10000.00   
5            0.0   5632.210000          5632.21          5000.00   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                0.00        0.00                     0.00   
1         435.17                0.00      117.08                     1.11   
2         605.67                0.00        0.00                     0.00   
3        2214.92               16.97        0.00                     0.00   
5         632.21                0.00        0.00                     0.00   

   last_pymnt_amnt  policy_code application_type  acc_now_delinq  delinq_amnt  
0           171.62            1       INDIVIDUAL               0            0  
1           119.66            1       INDIVIDUAL               0            0  
2           649.91            1       INDIVIDUAL               0            0  
3           357.48            1       INDIVIDUAL               0            0  
5           161.03            1       INDIVIDUAL               0            0

In [67]:
# Identify columns with constant values

constant_columns = df_for_loan.columns[df_for_loan.nunique() <= 1].tolist()

print("Constant columns to drop:", constant_columns)

Constant columns to drop: ['pymnt_plan', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'policy_code', 'application_type', 'acc_now_delinq', 'delinq_amnt']


In [68]:
# Dropping the columns which are having the same value
df_for_loan.drop(columns=constant_columns, inplace=True)

In [69]:
df_for_loan.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
5  1075269    1311441       5000         5000           5000.0   36 months   

  int_rate  installment grade sub_grade home_ownership  annual_inc  \
0   10.65%       162.87     B        B2           RENT     24000.0   
1   15.27%        59.83     C        C4           RENT     30000.0   
2   15.96%        84.33     C        C5           RENT     12252.0   
3   13.49%       339.31     C        C1           RENT     49200.0   
5    7.90%       156.46     A        A4           RENT     36000.0   

  verification_status issue_d  loan_status  \
0            Verified  Dec-11   Fully Paid   
1     Source Verified  Dec-11  Charged Off   
2        Not Verified  Dec-11   Fully Paid   
3     Source Verified  Dec-11   Fully Paid   
5     Source Verified  Dec-11   Fully Paid   

                                                 url         purpose zip_code  \
0  https://lendingclub.com/browse/loanDetail.acti...     credit_card    860xx   
1  https://lendingclub.com/browse/loanDetail.acti...             car    309xx   
2  https://lendingclub.com/browse/loanDetail.acti...  small_business    606xx   
3  https://lendingclub.com/browse/loanDetail.acti...           other    917xx   
5  https://lendingclub.com/browse/loanDetail.acti...         wedding    852xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  \
0         AZ  27.65            0           Jan-85               1         3   
1         GA   1.00            0           Apr-99               5         3   
2         IL   8.72            0           Nov-01               2         2   
3         CA  20.00            0           Feb-96               1        10   
5         AZ  11.20            0           Nov-04               3         9   

   pub_rec  revol_bal  total_acc   total_pymnt  total_pymnt_inv  \
0        0      13648          9   5863.155187          5833.84   
1        0       1687          4   1008.710000          1008.71   
2        0       2956         10   3005.666844          3005.67   
3        0       5598         37  12231.890000         12231.89   
5        0       7963         12   5632.210000          5632.21   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          5000.00         863.16                0.00        0.00   
1           456.46         435.17                0.00      117.08   
2          2400.00         605.67                0.00        0.00   
3         10000.00        2214.92               16.97        0.00   
5          5000.00         632.21                0.00        0.00   

   collection_recovery_fee  last_pymnt_amnt  
0                     0.00           171.62  
1                     1.11           119.66  
2                     0.00           649.91  
3                     0.00           357.48  
5                     0.00           161.03

In [70]:
df_for_loan.shape

(38577, 35)